### PR TITLE
feat: implement postgres query builder

### PR DESCRIPTION
**Pull request by hvn2k1**
## Summary

Part of #2306

This PR introduces a dialect-agnostic database client abstraction and a PostgreSQL query builder, enabling Bucketeer to support both MySQL and PostgreSQL backends transparently.

### Changes

- Add a unified `Client` interface with adapters for MySQL and PostgreSQL that abstracts transaction handling (`pkg/storage/v2/database/client.go`)
- Add pagination constants (`QueryNoLimit`, `QueryNoOffset`) for unified list operations (`pkg/storage/v2/database/constants.go`)
- Implement PostgreSQL query builder (`pkg/storage/v2/postgres/query.go`)
- Add mockgen directive and generated mock client for PostgreSQL (`pkg/storage/v2/postgres/`)
- Add unit tests for the database client and query builder

### CODE REVIEW

File: pkg/storage/v2/postgres/query.go [hunk 1]
```diff
@@ -25,3 +29,368 @@ func WritePlaceHolder(template string, start, count int) string {
...
+			return "", nil, next
+		}
+		sql = fmt.Sprintf("(%s::jsonb @> $%d::jsonb)", f.Column, next)
+		args = append(args, string(b))
+		return sql, args, next + 1
+	}
+	sql = fmt.Sprintf("%s %s $%d", f.Column, operatorMap[f.Operator], next)
+	args = append(args, f.Value)
+	return sql, args, next + 1
+}
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go:90]: Filter no longer emits IN/NOT IN; use InFilter / NotInFilter + tests. https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/database/client_test.go [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+	t.Run("propagates begin error", func(t *testing.T) {
+		t.Parallel()
+		ctrl := gomock.NewController(t)
+		pc := pgmock.NewMockClient(ctrl)
+		pc.EXPECT().RunInTransactionV2(gomock.Any(), gomock.Any()).Return(io.EOF)
+		db := NewPostgresStorageClient(pc)
+		err := db.RunInTransactionV2(ctx, func(context.Context) error { return nil })
+		assert.ErrorIs(t, err, io.EOF)
+	})
+}
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/database/client_test.go:141]: Added Postgres propagates callback error test (parity with MySQL). https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/postgres/query.go [hunk 3]
```diff
@@ -25,3 +29,368 @@ func WritePlaceHolder(template string, start, count int) string {
...
+		if i != 0 {
+			sb.WriteString(" AND ")
+		}
+		frag, fragArgs, after := wp.BindSQL(next)
+		sb.WriteString(frag)
+		args = append(args, fragArgs...)
+		next = after
+	}
+	sql = sb.String() + " "
+	return sql, args
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go:328]: Skip empty fragments; no stray AND; empty -> no WHERE. https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/postgres/query.go [hunk 4]
```diff
@@ -25,3 +29,368 @@ func WritePlaceHolder(template string, start, count int) string {
...
+	cur := next
+	for i, q := range f.Queries {
+		if i != 0 {
+			sb.WriteString(" OR ")
+		}
+		qs, qa, after := q.BindSQL(cur)
+		sb.WriteString(qs)
+		args = append(args, qa...)
+		cur = after
+	}
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go]: Skip empty sub-clauses; avoid invalid ( … OR ). https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/postgres/query.go [hunk 5]
```diff
@@ -25,3 +29,368 @@ func WritePlaceHolder(template string, start, count int) string {
...
+	case JSONLengthGreaterThan, JSONLengthSmallerThan:
+		if len(f.Values) == 0 {
+			return "", nil, next
+		}
+		op := ">"
+		if f.Func == JSONLengthSmallerThan {
+			op = "<"
+		}
+		sql = fmt.Sprintf("jsonb_array_length(%s::jsonb) %s %s", f.Column, op, f.Values[0])
+		return sql, nil, next
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go]: Length threshold bound as $n + args, not interpolated. https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/postgres/query.go [hunk 6]
```diff
@@ -25,3 +29,368 @@ func WritePlaceHolder(template string, start, count int) string {
...
+			for i, v := range f.Values {
+				if i != 0 {
+					sb.WriteString(", ")
+				}
+				sb.WriteString(fmt.Sprintf(`"%s"`, v))
+			}
+			sb.WriteString("]")
+		}
+		args = append(args, sb.String())
+		return sql, args, next + 1
```
[the remark this answers is not in the document]
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go]: SON payloads built with encoding/json (+ tests for edge cases). https://github.com/bucketeer-io/bucketeer/commit/ab0498f8ab5af6cc9941ce51bc7daa27cf96a132

File: pkg/storage/v2/postgres/query.go [hunk 7]
```diff
@@ -25,3 +29,445 @@ func WritePlaceHolder(template string, start, count int) string {
...
+		whereQuery, whereArgs := ConstructWhereSQLString(options.CreateWhereParts())
+		if whereArgs == nil {
+			whereArgs = []interface{}{}
+		}
+		return baseQuery + whereQuery, whereArgs
+	}
+	return baseQuery, []interface{}{}
+}
+
+type Orders struct {
```
Comment (t-kikuc) [inline pkg/storage/v2/postgres/query.go]: `mysql.Orders` also seems to have no callers in the repository and appears to be dead code too. 
If we are not removing `mysql.Orders` in this PR, please add a doc comment on this `postgres.Orders` struct stating it exists purely for parity with the MySQL counterpart.

File: pkg/storage/v2/database/client.go
Comment (t-kikuc) [inline pkg/storage/v2/database/client.go:1]: There is a SQL generator for Go called `squirrel`, which reduces the maintenance cost of this query builder code in mysql/postgres.

Let's state in the PR description why we implement it by ourselves instead of using it.

https://github.com/masterminds/squirrel

Comment (hvn2k1) [reply pkg/storage/v2/database/client.go:1]: Actually, the query builder for mysql is already here when I join, I don't really know why we choose it in the first place but maybe because of some complex query requirement or we want full control of the query builder.

Secondly, in the API layer we will abstract the `mysql.Client `into `database.Client` and `mysql.RunInTransaction` function into `database.RunInTransaction`. So maybe we'll need to switch mysql to use squirrel or sqlc or any query builder out there. Rewrite the whole storage layer is risky to me, so what I did is only minor refactoring in the API layer without touching the mysql storage implementation and implement Postgres alongside.

File: pkg/storage/v2/postgres/query.go (hunk 7 above)
Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go]: I remove it, thank you!

File: pkg/storage/v2/postgres/query.go [hunk 8]
```diff
@@ -25,3 +29,441 @@ func WritePlaceHolder(template string, start, count int) string {
...
+			whereArgs = []interface{}{}
+		}
+		return baseQuery + whereQuery, whereArgs
+	}
+	return baseQuery, []interface{}{}
+}
+
+const (
+	QueryNoLimit  = 0
+	QueryNoOffset = 0
```
Comment (Ubisoft-potato) [inline pkg/storage/v2/postgres/query.go]: These constants duplicate `database.QueryNoLimit` and `database.QueryNoOffset` defined in `pkg/storage/v2/database/constants.go`, which are never imported or used anywhere in the codebase.

Comment (hvn2k1) [reply pkg/storage/v2/postgres/query.go]: :+1: refactored https://github.com/bucketeer-io/bucketeer/pull/2540/changes/54222c95aaa26c35b55e375134d7d9faaf538f1e

**Review by cre8ivejp — APPROVED** [review]
Nice! Thanks!

**Review by Ubisoft-potato — APPROVED** [review]
LGTM! 👍
